File: .NET/Microsoft.Recognizers.Text.DataTypes.DataDrivenTests/TestTimexFormat.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Microsoft.Recognizers.Text.DataTypes.TimexExpression.Tests
{
    [TestClass]
    public class TestTimexFormat
    {
        [TestMethod]
        public void DataTypes_Format_Date()
        {
            Assert.AreEqual("2017-09-27", new TimexProperty { Year = 2017, Month = 9, DayOfMonth = 27 }.TimexValue);
            Assert.AreEqual("XXXX-WXX-3", new TimexProperty { DayOfWeek = 3 }.TimexValue);
            Assert.AreEqual("XXXX-12-05", new TimexProperty { Month = 12, DayOfMonth = 5 }.TimexValue);
        }

        [TestMethod]
        public void DataTypes_Format_Time()
        {
            Assert.AreEqual("T17:30:45", new TimexProperty { Hour = 17, Minute = 30, Second = 45 }.TimexValue);
            Assert.AreEqual("T05:06:07", new TimexProperty { Hour = 5, Minute = 6, Second = 7 }.TimexValue);
            Assert.AreEqual("T17:30", new TimexProperty { Hour = 17, Minute = 30, Second = 0 }.TimexValue);
            Assert.AreEqual("T23", new TimexProperty { Hour = 23, Minute = 0, Second = 0 }.TimexValue);
        }

        [TestMethod]
        public void DataTypes_Format_Duration()
        {
            Assert.AreEqual("P50Y", new TimexProperty { Years = 50 }.TimexValue);
            Assert.AreEqual("P6M", new TimexProperty { Months = 6 }.TimexValue);
            Assert.AreEqual("P3W", new TimexProperty { Weeks = 3 }.TimexValue);
            Assert.AreEqual("P5D", new TimexProperty { Days = 5 }.TimexValue);
            Assert.AreEqual("PT16H", new TimexProperty { Hours = 16 }.TimexValue);
            Assert.AreEqual("PT32M", new TimexProperty { Minutes = 32 }.TimexValue);
            Assert.AreEqual("PT20S", new TimexProperty { Seconds = 20 }.TimexValue);
        }

        [TestMethod]
        public void DataTypes_Format_Present()
        {
            Assert.AreEqual("PRESENT_REF", new TimexProperty { Now = true }.TimexValue);
        }

        [TestMethod]
        public void DataTypes_Format_DateTime()
        {
            Assert.AreEqual("XXXX-WXX-3T04", new TimexProperty { DayOfWeek = 3, Hour = 4, Minute = 0, Second = 0 }.TimexValue);
            Assert.AreEqual("2017-09-27T11:41:30", new TimexProperty { Year = 2017, Month = 9, DayOfMonth = 27, Hour = 11, Minute = 41, Second = 30 }.TimexValue);
        }

        [TestMethod]
        public void DataTypes_Format_DateRange()
        {
            Assert.AreEqual("2017", new TimexProperty { Year = 2017 }.TimexValue);
            Assert.AreEqual("SU", new TimexProperty { Season = "SU" }.TimexValue);
            Assert.AreEqual("2017-WI", new TimexProperty { Year = 2017, Season = "WI" }.TimexValue);
            Assert.AreEqual("2017-09", new TimexProperty { Year = 2017, Month = 9 }.TimexValue);
            Assert.AreEqual("2017-W37", new TimexProperty { Year = 2017, WeekOfYear = 37 }.TimexValue);
            Assert.AreEqual("2017-W37-WE", new TimexProperty { Year = 2017, WeekOfYear = 37, Weekend = true }.TimexValue);
            Assert.AreEqual("XXXX-05", new TimexProperty { Month = 5 }.TimexValue);
        }

        [TestMethod]
        public void DataTypes_Format_TimeRange()
        {
            Assert.AreEqual("TEV", new TimexProperty { PartOfDay = "EV" }.TimexValue);
        }

        [TestMethod]
        public void DataTypes_Format_DateTimeRange()
        {
            Assert.AreEqual("2017-09-27TEV", new TimexProperty { Year = 2017, Month = 9, DayOfMonth = 27, PartOfDay = "EV" }.TimexValue);
        }
    }
}

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/test/java/com/microsoft/recognizers/text/datatypes/timex/expression/TestTimexFormat.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.recognizers.text.datatypes.timex.expression;

import com.microsoft.recognizers.datatypes.timex.expression.TimexProperty;

import java.math.BigDecimal;

import org.junit.Assert;
import org.junit.Test;

public class TestTimexFormat {

    @Test
    public void dataTypesFormatDate() {
        Assert.assertEquals("2017-09-27", new TimexProperty() {
            {
                setYear(2017);
                setMonth(9);
                setDayOfMonth(27);
            }
        }.getTimexValue());
        Assert.assertEquals("XXXX-WXX-3", new TimexProperty() {
            {
                setDayOfWeek(3);
            }
        }.getTimexValue());
        Assert.assertEquals("XXXX-12-05", new TimexProperty() {
            {
                setMonth(12);
                setDayOfMonth(5);
            }
        }.getTimexValue());
    }

    @Test
    public void dataTypesFormatTime() {
        Assert.assertEquals("T17:30:45", new TimexProperty() {
            {
                setHour(17);
                setMinute(30);
                setSecond(45);
            }
        }.getTimexValue());
        Assert.assertEquals("T05:06:07", new TimexProperty() {
            {
                setHour(5);
                setMinute(6);
                setSecond(7);
            }
        }.getTimexValue());
        Assert.assertEquals("T17:30", new TimexProperty() {
            {
                setHour(17);
                setMinute(30);
                setSecond(0);
            }
        }.getTimexValue());
        Assert.assertEquals("T23", new TimexProperty() {
            {
                setHour(23);
                setMinute(0);
                setSecond(0);
            }
        }.getTimexValue());
    }

    @Test
    public void dataTypesFormatDuration() {
        Assert.assertEquals("P50Y", new TimexProperty() {
            {
                setYears(new BigDecimal(50));
            }
        }.getTimexValue());
        Assert.assertEquals("P6M", new TimexProperty() {
            {
                setMonths(new BigDecimal(6));
            }
        }.getTimexValue());
        Assert.assertEquals("P3W", new TimexProperty() {
            {
                setWeeks(new BigDecimal(3));
            }
        }.getTimexValue());
        Assert.assertEquals("P5D", new TimexProperty() {
            {
                setDays(new BigDecimal(5));
            }
        }.getTimexValue());
        Assert.assertEquals("PT16H", new TimexProperty() {
            {
                setHours(new BigDecimal(16));
            }
        }.getTimexValue());
        Assert.assertEquals("PT32M", new TimexProperty() {
            {
                setMinutes(new BigDecimal(32));
            }
        }.getTimexValue());
        Assert.assertEquals("PT20S", new TimexProperty() {
            {
                setSeconds(new BigDecimal(20));
            }
        }.getTimexValue());
    }

    @Test
    public void dataTypesFormatPresent() {
        Assert.assertEquals("PRESENT_REF", new TimexProperty() {
            {
                setNow(true);
            }
        }.getTimexValue());
    }

    @Test
    public void dataTypesFormatDateTime() {
        Assert.assertEquals("XXXX-WXX-3T04", new TimexProperty() {
            {
                setDayOfWeek(3);
                setHour(4);
                setMinute(0);
                setSecond(0);
            }
        }.getTimexValue());
        Assert.assertEquals("2017-09-27T11:41:30", new TimexProperty() {
            {
                setYear(2017);
                setMonth(9);
                setDayOfMonth(27);
                setHour(11);
                setMinute(41);
                setSecond(30);
            }
        }.getTimexValue());
    }

    @Test
    public void dataTypesFormatDateRange() {
        Assert.assertEquals("2017", new TimexProperty() {
            {
                setYear(2017);
            }
        }.getTimexValue());
        Assert.assertEquals("SU", new TimexProperty() {
            {
                setSeason("SU");
            }
        }.getTimexValue());
        Assert.assertEquals("2017-WI", new TimexProperty() {
            {
                setYear(2017);
                setSeason("WI");
            }
        }.getTimexValue());
        Assert.assertEquals("2017-09", new TimexProperty() {
            {
                setYear(2017);
                setMonth(9);
            }
        }.getTimexValue());
        Assert.assertEquals("2017-W37", new TimexProperty() {
            {
                setYear(2017);
                setWeekOfYear(37);
            }
        }.getTimexValue());
        Assert.assertEquals("2017-W37-WE", new TimexProperty() {
            {
                setYear(2017);
                setWeekOfYear(37);
                setWeekend(true);
            }
        }.getTimexValue());
        Assert.assertEquals("XXXX-05", new TimexProperty() {
            {
                setMonth(5);
            }
        }.getTimexValue());
    }

    @Test
    public void dataTypesFormatTimeRange() {
        Assert.assertEquals("TEV", new TimexProperty() {
            {
                setPartOfDay("EV");
            }
        }.getTimexValue());
    }

    @Test
    public void dataTypesFormatDateTimeRange() {
        Assert.assertEquals("2017-09-27TEV", new TimexProperty() {
            {
                setYear(2017);
                setMonth(9);
                setDayOfMonth(27);
                setPartOfDay("EV");
            }
        }.getTimexValue());
    }
}
